django/utils/functional.py

```diff
@@ -416,7 +423,23 @@ def __init__(self, func):
         super().__init__()
 
     def _setup(self):
-        self._wrapped = self._setupfunc()
+        if iscoroutinefunction(self._setupfunc):
+            self._wrapped = self._setupfunc
+        else:
+            self._wrapped = self._setupfunc()
+
+    def __await__(self):
+        async def _():
+            if iscoroutinefunction(self._wrapped):
+                self._wrapped = await self._wrapped()
+            elif callable(self._wrapped):
+                self._wrapped = await sync_to_async(self._wrapped)()
+            return self._wrapped
```

Comment:
I would love to see a set of comments on this code explaining exactly what it's doing and why - it took long enough to work out at the sprints how this was going to work that we should save others the trouble.

Reply:
I've added some comments as requested.